src/Algorithms/RFTrainer.cpp

```diff
@@ -238,385 +241,181 @@ void RFTrainer::train(RFClassifier& model, ClassificationDataset const& dataset)
-
-		//Index of attributes
-		std::size_t bestAttributeIndex, bestAttributeValIndex;
-
-		//Attribute values
-		double bestAttributeVal;
-		double bestImpurity = n+1.0;
-
-		for (set<std::size_t>::iterator it=tableIndicies.begin() ; it != tableIndicies.end(); it++ ){
-			std::size_t attributeIndex = *it;
-			boost::unordered_map<std::size_t, std::size_t> cTmpAbove = cAbove;
-			cBelow.clear();
-			for(std::size_t i=1; i<n; i++){
-				std::size_t prev = i-1;
-
-				//Update the count of the label
-				cBelow[dataset.element(tables[attributeIndex][prev].id).label]++;
-				cTmpAbove[dataset.element(tables[attributeIndex][prev].id).label]--;
-
-				if(tables[attributeIndex][prev].value!=tables[attributeIndex][i].value){
-					//n1 = Number of cases to the left child node
-					//n2 = number of cases to the right child node
-					std::size_t n1 = i;
-					std::size_t n2 = n-n1;
-
-					//Calculate the Gini impurity of the split
-					double impurity = n1*gini(cBelow,n1)+n2*gini(cTmpAbove,n2);
-					if(impurity<bestImpurity){
-						//Found a more pure split, store the attribute index and value
-						bestImpurity = impurity;
-						bestAttributeIndex = attributeIndex;
-						bestAttributeValIndex = prev;
-						bestAttributeVal = tables[attributeIndex][bestAttributeValIndex].value;
-						cBestAbove = cTmpAbove;
-						cBestBelow = cBelow;
-					}
-				}
-			}
-		}
-
-		if(bestImpurity<n+1){
-			AttributeTables rTables, lTables;
-			splitAttributeTables(tables, bestAttributeIndex, bestAttributeValIndex, lTables, rTables);
-			tables.clear();
-			//Continue recursively
-
-			nodeInfo.attributeIndex = bestAttributeIndex;
-			nodeInfo.attributeValue = bestAttributeVal;
-			nodeInfo.leftNodeId = 2*nodeId+1;
-			nodeInfo.rightNodeId = 2*nodeId+2;
-
-			lTree = buildTree(lTables, dataset, cBestBelow, nodeInfo.leftNodeId);
-			rTree = buildTree(rTables, dataset, cBestAbove, nodeInfo.rightNodeId);
-		}else{
-			//Leaf node
-			isLeaf = true;
-		}
+	std::size_t n = tables.size().n_elements;
 
+	if(detail::cart::gini(cFull,n)==0.0 || n <= m_nodeSize) {
+		nodeInfo.label = detail::cart::hist(cFull);
+		return tree;
 	}
 
-	//Store entry in the tree table
-	CARTClassifier<RealVector>::TreeType tree;
+	//Randomly select the attributes to test for split
+	auto tableIndices = generateRandomTableIndices(rng);
 
-	if(isLeaf){
-		nodeInfo.label = hist(cAbove);
-		tree.push_back(nodeInfo);
+	auto split = findSplit(tables,elements,cFull,tableIndices);
+
+	// if the purity hasn't improved, this is a leaf.
+	if(!split) {
+		// this shouldn't really happen: leaves ought to be cought by previous check
+		// TODO(jwrigley): Why is hist only applied to leaves, when average is applied to all nodes
+		nodeInfo.label = detail::cart::hist(cFull);
 		return tree;
 	}
+	nodeInfo <<= split;
+	auto lrTables = tables.split(split.splitAttribute, split.splitRow);
+	//Continue recursively
+
+	nodeInfo.leftNodeId  = (nodeId<<1)+1;   nodeInfo.rightNodeId = (nodeId<<1)+2;
```

Comment:
this is going to fail with trees that for some reason are very deep. We had this in the CARTTrainer. the logic should be after building the tree:

nodeInfo.leftNodeId = nodeId+1;
...
nodeInfo.rightNodeId = nodeInfo.leftNodeId + lTree.size();

This also happens in the other case